File: ocaml/allergies/allergies.ml
open Base

type allergen = Eggs | Peanuts | Shellfish | Strawberries | Tomatoes | Chocolate | Pollen | Cats

(* land operator is a bitwise and operator *)
let allergic_to score allergen =
  let ascore =
    match allergen with
    | Eggs -> 1
    | Peanuts -> 2
    | Shellfish -> 4
    | Strawberries -> 8
    | Tomatoes -> 16
    | Chocolate -> 32
    | Pollen -> 64
    | Cats -> 128
  in
  let result = score land ascore in
  result = ascore

let allergies score =
  [ Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats ]
  |> List.filter ~f:(allergic_to score)

File: go/allergies/allergies.go
package allergies

var allergenTable = map[string]uint{
	"eggs": 1,
	"peanuts": 2,
	"shellfish": 4,
	"strawberries": 8,
	"tomatoes": 16,
	"chocolate": 32,
	"pollen": 64,
	"cats": 128,
}

func Allergies(allergies uint) (acc []string) {
	for k, _ := range allergenTable {
		if AllergicTo(allergies, k) {
			acc = append(acc, k)
		}
	}
	return
}

func AllergicTo(allergies uint, allergen string) bool {
	v, ok := allergenTable[allergen]

	if !ok { return false }
	return v == (allergies & v)
}
